### PR TITLE
Add external loopback

### PR DESCRIPTION
**Pull request by RavidRothfeld**
Added external loopback function
Added ssh connectivtiy test function
Added results function

### CODE REVIEW

File: ngc_rdma_wrapper.sh [hunk 1]
```diff
@@ -15,46 +19,51 @@ help() {
...
     # Loop over the Host devices
     for i in "${!HOST_MLNX[@]}"; do
         # Store the current element's value in a variable
         element="${HOST_MLNX[i]}"
 
         # Seperate the element to PCI_DEVICE and MLX
-        PCI_DEVICE=$(echo "${element}" | cut -d ' ' -f 1)
-        MLX=$(echo "${element}" | cut -d ' ' -f 2)
+        PCI_DEVICE="$(echo "${element}" | cut -d ' ' -f 1)"
+        MLX="$(echo "${element}" | cut -d ' ' -f 2)"
```
Comment (blochl) [inline ngc_rdma_wrapper.sh:55]: This is a good change, but it has nothing to do with the functionality. So it should be in a separate commit.

File: ngc_rdma_wrapper.sh [hunk 2]
```diff
@@ -15,46 +19,51 @@ help() {
...
+        [[ "${PCI_DEVICE}" != *.1 ]] || continue
 
         # Store PCI Prefix & Set dual_port to false
         PCI_PREFIX="${element%.*}"
         dual_port=false
 
         # Loop and check if the current device is Dual Port
         for device in "${HOST_MLNX[@]}"; do
-            if [[ ${device} == "${PCI_PREFIX}"*".1 "* ]]; then
+            if [[ "${device}" == "${PCI_PREFIX}"*".1 "* ]]; then
```
Comment (blochl) [inline ngc_rdma_wrapper.sh:66]: This is a good change, but it has nothing to do with the functionality. So it should be in a separate commit.

File: ngc_rdma_wrapper.sh [hunk 3]
```diff
@@ -63,30 +72,191 @@ ngc_rdma_test() {
         # Store  the next index in the HOST_MLNX
         next_index=$((i + 1))
         # Determine whether there is another element in the HOST_MLNX to skip to
-        if [ ${next_index} -lt ${#HOST_MLNX[@]} ]; then
+        if [ "${next_index}" -lt "${#HOST_MLNX[@]}" ]; then
```
Comment (blochl) [inline ngc_rdma_wrapper.sh:75]: This is a good change, but it has nothing to do with the functionality. So it should be in a separate commit.

File: ngc_rdma_wrapper.sh [hunk 4]
```diff
@@ -63,30 +72,191 @@ ngc_rdma_test() {
...
+            echo -e "${WHITE}Dual Port -  1st Port: ${MLX} PCI: ${PCI_DEVICE} | 2nd Port: ${MLX2} PCI: ${PCI_DEVICE2}${NC}" &>> "${LOGFILE}"
+            "${scriptdir}/ngc_rdma_test.sh" "${HOST_IP}" "${MLX}","${MLX2}" "${CLIENT_IP}" "${MLX}","${MLX2}" ${use_cuda} &>> "${LOGFILE}"
 
-            # If the device is Single Port:
+        # If the device is Single Port:
         else
-            echo -e "${WHITE}Single Port - ${MLX} Located on PCI: ${PCI_DEVICE}${NC}"
-            "${scriptdir}/ngc_rdma_test.sh" "${HOST_IP}" "${MLX}" "${CLIENT_IP}" "${MLX}" ${use_cuda}
+            echo -e "${WHITE}Single Port - ${MLX} Located on PCI: ${PCI_DEVICE}${NC}" &>> "${LOGFILE}"
+            "${scriptdir}/ngc_rdma_test.sh" "${HOST_IP}" "${MLX}" "${CLIENT_IP}" "${MLX}" ${use_cuda} &>> "${LOGFILE}"
```
Comment (blochl) [inline ngc_rdma_wrapper.sh:91]: Using a logfile is great, but it has nothing to do with the declared functionality of this commit. So it deserves a separate commit.

File: ngc_rdma_wrapper.sh [hunk 5]
```diff
@@ -2,11 +2,15 @@
...
-CLIENT_IP=${1}
-HOST_IP=${2}
+CLIENT_IP=${2}
+HOST_IP=${1}
 WHITE='\033[1;37m'
+GREEN='\033[1;32m'
+RED='\033[1;31m'
 NC='\033[0m'
 scriptdir="$(dirname "$0")"
+LOGFILE="/tmp/ngc-rdma-test_$(date +%H:%M:%S__%d-%m-%Y).log"
```
Comment (blochl) [inline ngc_rdma_wrapper.sh:12]: If it's a log file, it shouldn't be in `/tmp`. If it's a temporary utility file, you can use `mktemp`.

File: ngc_rdma_wrapper.sh [hunk 6]
```diff
@@ -63,30 +72,191 @@ ngc_rdma_test() {
...
+        COMBINED_ARRAY+=("${HOST_MLNX[j]} ${CABLE_ARRAY[j]}")
+    done
+
+    # Loop and check for similar Cables' SN, and send it to the script
+    for element in "${COMBINED_ARRAY[@]}"; do
+        # Skip iteration if one of the devices have already been iterated
+        if (( ${#processed_array[@]} != 0 )) &&
+            printf '%s\n' "${processed_array[@]}" | grep -q "^\(${another_element}\|${element}\|${dualmlx2}\|${dualmlx}\)\$"; then
+            continue
+        fi
```
Comment (blochl) [inline ngc_rdma_wrapper.sh:125]: This can be split to some `check_if_in_array` function.

File: ngc_rdma_wrapper.sh [hunk 7]
```diff
@@ -63,30 +72,191 @@ ngc_rdma_test() {
...
+        done
+    done
+}
+
+
+# Check for SSH connectivity
+check_ssh() {
+    local failed_hosts=""
+    for ip in "$@"; do
+        if ! ssh -q -oStrictHostKeyChecking=no "${ip}" exit; then
```
Comment (blochl) [inline ngc_rdma_wrapper.sh:201]: Probably should also use the `-o UserKnownHostsFile=/dev/null` option as well. Otherwise the `~/.ssh/known_hosts` file will be written to every time the script runs, and it will be super-long.

File: ngc_rdma_wrapper.sh [hunk 8]
```diff
@@ -63,30 +72,191 @@ ngc_rdma_test() {
...
+            failed_hosts+=" ${ip}"
+        fi
+    done
+    if [[ -n "${failed_hosts}" ]]; then
+        echo "SSH connection failed for:${failed_hosts}. Exiting.."
+        exit 1
+    else
+        echo "Created log file: ${LOGFILE}"
```
Comment (blochl) [inline ngc_rdma_wrapper.sh:209]: Why here?

**Review by blochl — APPROVED** [review]
Approving, despite the mentioned issues. @RavidRothfeld - please go over them. At least the issue with `known_hosts` file will need fixing in the future.

File: ngc_rdma_wrapper.sh (hunk 5 above)
Comment (RavidRothfeld) [reply ngc_rdma_wrapper.sh:12]: Leaving it under /tmp as synced

**Review by blochl — APPROVED** [review]
